File: libraries/bridgerton/tests/values.js
// Shared integration test in Node and Chromium. No runtime-specific dependencies.
async function testValues(counter) {
  const check = (condition, message) => { if (!condition) throw Error(message); };
  const equal = (a, b) => JSON.stringify(a) === JSON.stringify(b);
  const rejects = fn => { let rejected = false; try { fn(); } catch { rejected = true; } check(rejected, 'invalid input must throw'); };
  const beforeBadCall = counter.live_counters();
  rejects(() => counter.consume_with_card({}, counter.optional_object(true)));
  check(counter.live_counters() === beforeBadCall, 'failed value decoding releases an already-transferred object');
  rejects(() => counter.fail_value_output());
  let valueFailure;
  try { await counter.fail_value_output_later(); } catch (error) { valueFailure = String(error); }
  check(valueFailure.includes('intentional Tsify serialization failure'), 'Tsify conversion failure is recoverable');
  counter.value(); // The receiver borrow was released on both conversion failures.
  const card = counter.sample_card();
  const wire = { value: { kind: 'in_progress', payload: { completedCount: 3 } } };
  check(equal(counter.echo_wire_state(wire), wire), 'Tsify preserves explicit Serde tags, names, and generic records');
  check(equal(counter.echo_wire_state({ value: { kind: 'not_started' } }), { value: { kind: 'not_started' } }), 'unit variant tag');
  check(counter.echo_wire_id(42) === 42, 'only serde transparent unwraps a record');
  check(counter.maybe_wire_state(null) === undefined && counter.maybe_wire_state(undefined) === undefined, 'optional generic outputs are undefined');
  check(equal(counter.maybe_wire_state(wire), wire), 'optional generic value');

  rejects(() => counter.fail_encoding());
  let encodingError;
  try { await counter.fail_encoding_later(); } catch (error) { encodingError = String(error); }
  check(encodingError === 'intentional encoding failure', 'async output conversion error');
  // Node and Chromium later free this Counter, which detects a stranded borrow.
  check(equal(counter.echo_alias(card), card), 'aliased value conversion');
  check(equal(counter.maybe_card(card), card), 'optional value input/output');
  check(counter.maybe_card(null) == null && counter.maybe_card(undefined) == null, 'nullable and omitted value');
  check(equal(counter.cards, [card]), 'collection getter');
  check(equal(await counter.echo_nested([[null, card], []]), [[null, card], []]), 'async nested nullable collections');
  check(card.term.text === '語 🦀' && card.term.gloss === 'language', 'nested UTF-8 record');
  check(equal(card.alternatives, [{ text: '言葉' }]), 'optional record');
  check(card.state === 'New' && equal(card.tags, ['日本語', '']), 'enum and array');
  const state = { Learning: { steps: 4294967295, due: undefined } };
  const revised = counter.revise_card(card, state);
  check(revised.id === card.id + 1 && equal(revised.state, state) && revised.starred, 'two value inputs');
  check(card.state === 'New' && !card.starred, 'owned copy');
  check(equal(counter.echo_cards([null, revised, card]), [null, revised, card]), 'nested collections');
  check(equal(counter.echo_cards([]), []), 'empty collection');
  for (const state of ['New', { Learning: { steps: 0, due: 'tomorrow' } }, { Known: '語' }, { Pair: [7, false] }]) {
    check(equal(counter.echo_state(state), state), 'enum payload forms');
  }
  check(counter.rename_card(card, 'a\0b').term.text === 'a\0b', 'NUL and String input');
  counter.observe_three((number, text, flag) => check(number === 42 && text === 'three' && flag, 'three-argument callback'));
  const indices = counter.echo_indices(new Uint32Array([0, 42, 4294967295]));
  check(indices instanceof Uint32Array && indices[2] === 4294967295, 'numeric vector ABI stays a typed array');
  const bytes = counter.echo_bytes(new Uint8Array([0, 128, 255]));
  check(bytes instanceof Uint8Array && bytes.join(',') === '0,128,255', 'byte buffer ABI');
  const nestedBytes = counter.nested_bytes();
  check(nestedBytes.every(bytes => bytes instanceof Uint8Array) && nestedBytes[0][2] === 255 && nestedBytes[1].length === 0, 'nested byte buffers');
  const moved = counter.optional_object(true);
  const alias = moved;
  check(counter.consume(moved) === 0, 'owned object input');
  rejects(() => alias.value());
  check(counter.consume_optional(undefined) == null, 'absent object input');
  check(counter.consume_optional(counter.optional_object(true)) === 0, 'optional object input');
  check(await counter.consume_later(counter.optional_object(true)) === 0, 'async owned object input');
  let savedObject;
  counter.emit_object(object => { savedObject = object; counter.value(); });
  check(savedObject.value() === 0, 'callback object retained');
  check(counter.consume(savedObject) === 0, 'callback object consumed');
  counter.emit_objects((first, label, last) => {
    check(label === 'objects' && counter.consume(first) === 0, 'mixed object callback and reentry');
    savedObject = last;
  });
  check(savedObject.value() === 0, 'second callback object retained');
  savedObject.free();
  const objects = counter.object_list();
  check(objects.map(value => value.value()).join(',') === '1,2,3', 'object array');
  objects.forEach(value => value.free());
  const present = counter.optional_object(true);
  check(present.value() === 0 && counter.optional_object(false) == null, 'optional object');
  present.free();
  const nestedObjects = counter.nested_objects();
  check(nestedObjects.length === 2 && nestedObjects[0] == null && nestedObjects[1].value() === 0, 'nested object array');
  nestedObjects[1].free();
  const laterObjects = await counter.objects_later();
  check(laterObjects.length === 3, 'async object array');
  laterObjects.forEach(value => value.free());
  check(JSON.stringify(counter.nested_values()) === '[["one"],[],["two"]]', 'nested value arrays');

  const controller = new AbortController();
  const signal = controller.signal;
  let listeners = 0;
  const add = signal.addEventListener.bind(signal);
  const remove = signal.removeEventListener.bind(signal);
  signal.addEventListener = (...args) => { listeners++; return add(...args); };
  signal.removeEventListener = (...args) => { listeners--; return remove(...args); };
  check(await counter.abortable_wait(1, signal) === false, 'normal signal completion');
  check(listeners === 0, 'completion removes abort listener');
  const waits = [counter.abortable_wait(60000, signal), counter.abortable_wait(60000, signal)];
  await new Promise(resolve => setTimeout(resolve, 5));
  controller.abort();
  check((await Promise.all(waits)).every(Boolean), 'browser controller aborts all waiters');
  check(listeners === 0, 'abort removes listeners');
  check(await counter.abortable_wait(60000, signal), 'pre-aborted signal');
  check(await counter.abortable_wait(1) === false, 'optional signal');

  const token = new AbortController();
  const pending = counter.card_later(revised, token.signal);
  // wasm-bindgen starts async exports on a microtask. Wait for deserialization
  // and the first Rust suspension before testing independence from the JS object.
  await Promise.resolve();
  check(counter.active_operations() === 1, 'Rust data future has started');
  revised.term.text = 'changed after call';
  const later = await pending;
  check(later.term.text === '語 🦀' && equal(later.state, { Known: 'remembered' }), 'suspended future owns decoded data');
  const cancelled = new AbortController();
  cancelled.abort();
  let error;
  try { await counter.card_later(card, cancelled.signal); } catch (e) { error = e; }
  check(error instanceof Error && error.message === 'operation aborted', 'data cancellation');
  // Errors are Error objects; typed errors carry a detail shaped like Swift's.
  try { counter.io_result(); } catch (e) { error = e; }
  check(error instanceof Error && error.detail.kind === 'NotFound' && error.message.includes('missing test pack'), 'io error detail');
  try { counter.source_result(); } catch (e) { error = e; }
  check(error instanceof Error && error.detail.type === 'Read' && error.detail.value === 'opaque source', 'typed error detail');
  try { counter.typed_result(true); } catch (e) { error = e; }
  check(error instanceof Error && error.detail.InvalidAnswer.term === '語' && error.detail.InvalidAnswer.attempts === 2, 'value error detail');
  try { counter.string_error(); } catch (e) { error = e; }
  check(error instanceof Error && error.message === 'plain error' && error.detail === undefined, 'string error');
  rejects(() => counter.revise_card({ ...card, term: { ...card.term, text: '' } }, 'New'));
  rejects(() => counter.echo_state({ Unknown: {} }));
  rejects(() => counter.echo_state({ Pair: ['wrong', true] }));
  rejects(() => counter.echo_cards([{ ...card, id: -1 }]));
  rejects(() => counter.echo_cards([{ ...card, starred: 'true' }]));
  rejects(() => counter.echo_cards([{ ...card, term: {} }]));
  check(equal(counter.echo_cards([{ ...card, extraUIProperty: true }]), [card]), 'extra JS properties are ignored');

  return 'PASS: generated JS records/enums, nested arrays/options, sync/async roundtrips, owned inputs, and malformed values';
}
if (typeof module !== 'undefined') module.exports = { testValues };

File: libraries/bridgerton/tests/Values.swift
import Foundation
import BridgeFFI
import SwiftUI

// Compile a real SwiftUI view against the generated value types.
struct CardPreview: View {
    let card: Card
    var body: some View {
        VStack {
            Text(card.term.text)
            if let gloss = card.term.gloss { Text(gloss) }
            ForEach(card.alternatives.indices, id: \.self) { i in Text(card.alternatives[i].text) }
            switch card.state {
            case .New: Text("New")
            case .Learning(let steps, let due): Text("\(steps): \(due ?? "unscheduled")")
            case .Known(let note): Text(note)
            case .Pair(let count, let enabled): Text(verbatim: "\(count): \(enabled)")
            }
        }
    }
}

private func require(_ condition: Bool, _ message: String) {
    precondition(condition, message)
}
private func rejects(_ body: () throws -> Void) {
    do { try body(); fatalError("expected invalid value to be rejected") }
    catch is BridgeError {}
    catch { fatalError("unexpected error: \(error)") }
}

@MainActor func testValues(_ counter: Counter) async throws {
    let card = counter.sample_card()
    require(counter.echo_alias(card: card) == card, "aliased value conversion")
    require(counter.maybe_card(card: card) == card && counter.maybe_card(card: nil) == nil, "optional value input/output")
    require(counter.cards == [card], "collection getter")
    require(await counter.echo_nested(cards: [[nil, card], []]) == [[nil, card], []], "async nested nullable collections")
    require(card.term.text == "語 🦀" && card.term.gloss == "language", "nested UTF-8 record")
    require(card.alternatives == [Term(text: "言葉", gloss: nil)], "optional field")
    require(card.state == .New && card.tags == ["日本語", ""], "enum and array")
    let state = ReviewState.Learning(steps: UInt32.max, due: nil)
    let revised = try counter.revise_card(card: card, state: state)
    require(revised.id == card.id + 1 && revised.state == state && revised.starred, "two value inputs")
    require(card.state == .New && !card.starred, "value semantics preserve caller input")
    require(counter.echo_cards(cards: [nil, revised, card]) == [nil, revised, card], "nested collections")
    require(counter.echo_cards(cards: []) == [], "empty collection")
    for state in [ReviewState.New, .Learning(steps: 0, due: "tomorrow"), .Known("語"), .Pair(7, false)] {
        require(counter.echo_state(state: state) == state, "all enum payload forms")
    }
    let renamed = counter.rename_card(card: card, text: "a\0b")
    require(renamed.term.text == "a\0b", "embedded NUL and String input")
    let token = AbortController()
    let later = try await counter.card_later(card: revised, cancellation: token.signal())
    require(later.id == revised.id + 1 && later.state == .Known("remembered"), "async owned data")
    let cancelled = AbortController()
    cancelled.abort()
    do {
        _ = try await counter.card_later(card: revised, cancellation: cancelled.signal())
        fatalError("expected data future cancellation")
    } catch let error as BridgeError { require(error.description == "operation aborted", "data future cancellation") }
    var invalid = card
    invalid.term.text = ""
    rejects { _ = try counter.revise_card(card: invalid, state: .New) }
    // Plain data can cross actors without touching confined Rust objects.
    let copied = await Task.detached { revised }.value
    require(copied == revised, "generated data conforms to Sendable")
    _ = CardPreview(card: copied)
    try testCodec()
    let tree = Tree.Record(TreeRecord(child: .Next(.Pair(BridgePair(.Leaf(7), true)))))
    require(counter.echo_tree(tree: tree) == tree, "recursive enum through a record, optional, and pair")
    let mutual = MutualA.Next(.Next(.End))
    require(counter.echo_mutual(tree: mutual) == mutual, "mutually recursive enum layout")
    let chain = Chain_UInt32.Link(1, .Link(2, .End))
    require(counter.echo_chain(chain: chain) == chain, "generic recursive enum layout")
    let collection = CollectionTree.Map(["nested": .Children([.Set([.Leaf(3)])])])
    require(counter.echo_collection_tree(tree: collection) == collection, "recursive collections need no extra enum indirection")
    print("PASS: generated Swift records/enums, nested arrays/options, all enum forms, sync/async roundtrips, SwiftUI view, and malformed values")
}

@MainActor private func testCodec() throws {
    let value: [String]? = ["語"]
    var writer = BridgeWriter()
    try value.bridgeWrite(&writer)
    let expected: [UInt8] = [1, 0,0,0,1, 0,0,0,3, 0xe8,0xaa,0x9e]
    require(writer.bytes == expected, "shared wire-format golden vector")
    for end in 0..<expected.count {
        rejects {
            var reader = BridgeReader(data: Data(expected.prefix(end)))
            _ = try Optional<Array<String>>.bridgeRead(&reader)
        }
    }
    var mapWriter = BridgeWriter()
    let map: [UInt32: UInt16?] = [1: 2]
    try map.bridgeWrite(&mapWriter)
    require(mapWriter.bytes == [0,0,0,1, 0,0,0,1, 1,0,2], "map wire vector matches Rust")
    var mapReader = BridgeReader(data: Data(mapWriter.bytes))
    require(try Dictionary<UInt32, UInt16?>.bridgeRead(&mapReader) == map, "optional map values")
    rejects { var r = BridgeReader(data: Data([0,0,0,2, 0,0,0,1, 0, 0,0,0,1, 0])); _ = try Dictionary<UInt32, UInt16?>.bridgeRead(&r) }
    rejects { var r = BridgeReader(data: Data([0,0,0,2, 0,0,0,1, 0,0,0,1])); _ = try Set<UInt32>.bridgeRead(&r) }
    for bits: UInt64 in [0, 1 << 63, 0x7ff0000000000000, 0x7ff8000000001234] {
        var w = BridgeWriter()
        try Double(bitPattern: bits).bridgeWrite(&w)
        var r = BridgeReader(data: Data(w.bytes))
        require(try Double.bridgeRead(&r).bitPattern == bits, "floating-point bits survive exactly")
    }
    var pairWriter = BridgeWriter()
    try BridgePair(UInt8(1), Int16(-2)).bridgeWrite(&pairWriter)
    require(pairWriter.bytes == [1,255,254], "pair wire vector matches Rust")
    var timestampWriter = BridgeWriter()
    let timestamp = BridgeTimestamp(seconds: -1, nanoseconds: 123_456_789)
    try timestamp.bridgeWrite(&timestampWriter)
    require(timestampWriter.bytes == [255,255,255,255,255,255,255,255, 7,91,205,21], "timestamp wire vector matches Rust")
    rejects { var r = BridgeReader(data: Data([2])); _ = try Bool.bridgeRead(&r) }
    rejects { var r = BridgeReader(data: Data([2])); _ = try Optional<UInt32>.bridgeRead(&r) }
    rejects { var r = BridgeReader(data: Data([0,0,0,1,0xff])); _ = try String.bridgeRead(&r) }
    rejects { var r = BridgeReader(data: Data([0xff,0xff,0xff,0xff])); _ = try Array<UInt32>.bridgeRead(&r) }
    rejects { var r = BridgeReader(data: Data([0,0,0,99])); _ = try ReviewState.bridgeRead(&r) }
    rejects { var w = BridgeWriter(); try Array(repeating: false, count: bridgeMaxItems + 1).bridgeWrite(&w) }
    rejects { var w = BridgeWriter(); try Array(repeating: Array(repeating: false, count: 256), count: 256).bridgeWrite(&w) }
    rejects { var r = BridgeReader(data: Data(), depth: bridgeMaxDepth); _ = try r.nested { _ in 0 } }
    rejects { var w = BridgeWriter(depth: bridgeMaxDepth); _ = try w.nested { _ in 0 } }
}
